src/uct/tcp/tcp_sockcm_ep.c

```diff
@@ -701,6 +694,101 @@ static ucs_status_t uct_tcp_sockcm_ep_client_init(uct_tcp_sockcm_ep_t *cep,
     return status;
 }
 
+static ucs_status_t uct_tcp_sockcm_ep_server_create(uct_tcp_sockcm_ep_t *tcp_ep,
+                                                    const uct_ep_params_t *params,
+                                                    uct_ep_h *ep_p)
+{
+    uct_tcp_sockcm_t *tcp_sockcm = uct_tcp_sockcm_ep_get_cm(tcp_ep);
+    uct_tcp_sockcm_t *params_tcp_sockcm;
+    ucs_async_context_t *new_async_ctx;
+    ucs_status_t status;
+
+    if (!(params->field_mask & UCT_EP_PARAM_FIELD_CM)) {
+        ucs_error("UCT_EP_PARAM_FIELD_CM is not set. field_mask 0x%lx",
+                  params->field_mask);
+        status = UCS_ERR_INVALID_PARAM;
+        goto err;
+    }
+
+    if (params->cm == NULL) {
+        ucs_error("cm cannot be NULL (ep=%p fd=%d)", tcp_ep, tcp_ep->fd);
+        status = UCS_ERR_INVALID_PARAM;
+        goto err;
+    }
+
+    params_tcp_sockcm = ucs_derived_of(params->cm, uct_tcp_sockcm_t);
+
+    /* check if the server opened this ep, to the client, on a CM that is
+     * different from the one it created its internal ep on earlier, when it
+     * received the connection request from the client (the cm used by its listener) */
+    if (&tcp_sockcm->super != params->cm) {
+        status = ucs_async_remove_handler(tcp_ep->fd, 1);
+        if (status != UCS_OK) {
+            ucs_error("failed to remove fd %d from the async handlers: %s",
+                      tcp_ep->fd, ucs_status_string(status));
+            goto err_unblock;
+        }
+    }
+
+    UCS_ASYNC_BLOCK(tcp_sockcm->super.iface.worker->async);
+
+    UCS_CLASS_CLEANUP(uct_cm_base_ep_t, &tcp_ep->super);
+
+    /* set the server's ep to use the cm from params and its iface
+     * (it could be the previous one it had - the one used by the listener or
+     * a new one set by the user) */
+    status = UCS_CLASS_INIT(uct_cm_base_ep_t, &tcp_ep->super, params);
+    if (status != UCS_OK) {
+        ucs_error("failed to initialize a uct_cm_base_ep_t endpoint");
+        goto err_unblock;
+    }
+
+    ucs_assert(uct_tcp_sockcm_ep_get_cm(tcp_ep) == params_tcp_sockcm);
+
+    status = UCT_CM_SET_CB(params, UCT_EP_PARAM_FIELD_SOCKADDR_NOTIFY_CB_SERVER,
+                           tcp_ep->super.server.notify_cb, params->sockaddr_cb_server,
+                           uct_cm_ep_server_conn_notify_callback_t,
+                           ucs_empty_function);
+    if (status != UCS_OK) {
+        goto err_unblock;
```

Comment:
add UCS_CLASS_CLEANUP to error flow

Reply:
but if there is an error then the dtor of tcp_ep will be called (from uct_tcp_sockcm_ep_create) and destroy its current cm_base ep

Reply:
where is it called?

Reply:
in line 841 
https://github.com/openucx/ucx/pull/5324/files#diff-0a7c42d102df8d010e7a9bec845edf0dR841